src/langchain_google_spanner/document_loader.py

```diff
@@ -0,0 +1,401 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import json
+from dataclasses import dataclass
+from typing import Any, Dict, Iterator, List, Union
+
+from google.cloud.spanner import Client, KeySet  # type: ignore
+from google.cloud.spanner_admin_database_v1.types import DatabaseDialect  # type: ignore
+from google.cloud.spanner_v1.data_types import JsonObject  # type: ignore
+from langchain_community.document_loaders.base import BaseLoader
+from langchain_core.documents import Document
+
+OPERATION_TIMEOUT_SECONDS = 240
+MUTATION_BATCH_SIZE = 1000
+
+CONTENT_COL_NAME = "page_content"
+METADATA_COL_NAME = "langchain_metadata"
+
+
+@dataclass
+class Column:
+    name: str
+    data_type: str
+    nullable: bool = True
+
+
+def _load_row_to_doc(
+    format: str,
+    content_columns: List[str],
+    metadata_columns: List[str],
+    metadata_json_column: str,
+    row: Dict,
+) -> Document:
+    page_content = ""
+    if format == "text":
+        page_content = " ".join(str(row[c]) for c in content_columns if c in row)
+    elif format == "YAML":
+        page_content = "\n".join(
+            f"{c}: {str(row[c])}" for c in content_columns if c in row
+        )
+    elif format == "JSON":
+        j = {}
+        for c in content_columns:
+            if c in row:
+                j[c] = row[c]
+        page_content = json.dumps(j)
+    elif format == "CSV":
+        page_content = ", ".join(str(row[c]) for c in content_columns if c in row)
+    if not page_content:
+        raise Exception("column page_content doesn't exist.")
+
+    metadata: Dict[str, Any] = {}
+    if metadata_json_column in metadata_columns and row.get(metadata_json_column):
+        metadata = row[metadata_json_column]
+
+    for c in metadata_columns:
+        if c in row and c != metadata_json_column:
+            metadata[c] = row[c]
+
+    return Document(page_content=page_content, metadata=metadata)
+
+
+def _load_doc_to_row(
+    table_fields: List[str],
+    doc: Document,
+    metadata_json_column: str,
+    parse_json: bool = True,
+):
+    """
+    Load document to row.
+
+    Args:
+        table_fields: Spanner table fields names.
+        doc: Document that is used.
+        metadata_json_column: Name of the special JSON column.
+        parse_json: Parse json column to string or leave it as JSON object. String format is needed to for Spanner inserts.
+                    JSON object is used to compare with Spanner reads.
+    """
+    doc_metadata = doc.metadata.copy()
+    row = [doc.page_content]
```

Comment:
I don't think we can automatically set this as the first field in the table, right? The content column could be anywhere in the list of fields, right?

Reply:
Ohhhh, thanks for flagging this. Updated to check with the columns arrangement. As long as it matches the fields during batch.insert then it should be good. p/s: batch.insert takes in `columns` and `values`.